[BUG][JAVA] Implement adapter for java.time.LocalDateTime in client code

#### Bug Report Checklist

- [x] Have you provided a full/minimal spec to reproduce the issue?
- [x] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [x] Have you [tested with the latest master](https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-to-test-with-the-latest-master-of-openapi-generator) to confirm the issue still exists?
- [x] Have you searched for related issues/PRs?
- [ ] What's the actual output vs expected output?
- [ ] [Optional] Sponsorship to speed up the bug fix or feature request ([example](https://github.com/OpenAPITools/openapi-generator/issues/6178))

<!--
Please follow the issue template below for bug reports.
Also please indicate in the issue title which language/library is concerned. Eg:  [BUG][JAVA] Bug generating foo with bar 
-->

##### Description
In the Java generator there is no adapter for `java.time.LocalDateTime` parsing in the client code generator. Adapters for `OffsetDateTime` and `LocalDate` exist, though. When Receiving a LocalDateTime over the API, the generated client tries to parse an object like this: 
```json
{
    "year" : 2010,
    "month" : "JANUARY",
    "dayOfMonth" : 1,
    "dayOfWeek" : "FRIDAY",
    "dayOfYear" : 1,
    "monthValue" : 1,
    "hour" : 2,
    "minute" : 2,
    "second" : 0,
    "nano" : 0,
    "chronology" : {
      "id" : "ISO",
      "calendarType" : "iso8601"
}
```
Because it receives a ISO string in this format `2010-01-01T02:02:00` it can't be parsed so an IllegalArgumentException like this is thrown:
```
Caused by: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was STRING at path $.letzteAenderung
	at com.google.gson.internal.bind.JsonTreeReader.expect(JsonTreeReader.java:163)
	at com.google.gson.internal.bind.JsonTreeReader.beginObject(JsonTreeReader.java:88)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:214)
	... 51 more
```

With a corresponding adapter for `LocalDateTime` parsing, I would expect to be able to parse a valid ISO_LOCAL_DATE_TIME string into a `LocalDateTime` object. 

<!-- describe what is the question, suggestion or issue and why this is a problem for you. -->

##### openapi-generator version
6.2.1, but it seems to be the case in earlier versions, too
<!-- which version of openapi-generator are you using, is it a regression? -->

##### OpenAPI declaration file content or url

<!-- if it is a bug, a json or yaml that produces it.
If you post the code inline, please wrap it with
```yaml
(here your code)
```
(for YAML code) or
```json
(here your code)
```
(for JSON code), so it becomes more readable. If it is longer than about ten lines,
please create a Gist (https://gist.github.com) or upload it somewhere else and
link it here.
  -->

Minimized excerpt from pet store example:
```yml
openapi: 3.0.3
info:
  title: Swagger Petstore - OpenAPI 3.0
  description: ''
  version: 0.0.1
tags:
  - name: pet
    description: Everything about your Pets
    externalDocs:
      description: Find out more
      url: http://swagger.io
  - name: store
    description: Access to Petstore orders
    externalDocs:
      description: Find out more about our store
      url: http://swagger.io
  - name: user
    description: Operations about user
paths:
  /store/order/{orderId}:
    get:
      tags:
        - store
      summary: Find purchase order by ID
      description: For valid response try integer IDs with value <= 5 or > 10. Other values will generate exceptions.
      operationId: getOrderById
      parameters:
        - name: orderId
          in: path
          description: ID of order that needs to be fetched
          required: true
          schema:
            type: integer
            format: int64
      responses:
        '200':
          description: successful operation
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Order'          
            application/xml:
              schema:
                $ref: '#/components/schemas/Order'
        '400':
          description: Invalid ID supplied
        '404':
          description: Order not found
components:
  schemas:
    Order:
      type: object
      properties:
        id:
          type: integer
          format: int64
          example: 10
        petId:
          type: integer
          format: int64
          example: 198772
        quantity:
          type: integer
          format: int32
          example: 7
        shipDate:
          type: string
          format: date-time
        status:
          type: string
          description: Order Status
          example: approved
          enum:
            - placed
            - approved
            - delivered
        complete:
          type: boolean
      xml:
        name: order

```

##### Generation Details

<!-- 
    Prefer CLI steps, including the language, libraries and various options. 
    Providing config-based settings allows for simpler testing across CLI and plugins. 
    For examples, see https://github.com/OpenAPITools/openapi-generator/tree/master/bin/configs
-->
Maven plugin configuration:
```xml
<plugin>
    <groupId>org.openapitools</groupId>
    <artifactId>openapi-generator-maven-plugin</artifactId>
    <version>6.2.1</version>
    <executions>
        <execution>
            <phase>package</phase>
            <goals>
                <goal>generate</goal>
            </goals>
            <configuration>
                <inputSpec>${project.basedir}/target/${project.artifactId}/openapi.yaml</inputSpec>
                <output>${project.build.directory}/generated-sources/openapi</output>
                <generatorName>java</generatorName>
                <groupId>${project.groupId}</groupId>
                <artifactId>${project.artifactId}</artifactId>
                <artifactVersion>${project.version}</artifactVersion>
                <additionalProperties>
                    serializableModel=true,dateLibrary=java8,java8=true,useTags=true,groupId=de.persona.dal,hideGenerationTimestamp=true
                </additionalProperties>
                <configOptions>
                    <sourceFolder>src/main/java</sourceFolder>
                </configOptions>
                <typeMappings>
                    <typeMapping>OffsetDateTime=LocalDateTime</typeMapping>
                </typeMappings>
                <importMappings>
                    <importMapping>java.time.OffsetDateTime=java.time.LocalDateTime</importMapping>
                </importMappings>
                <generateModelTests>false</generateModelTests>
                <generateApiTests>false</generateApiTests>
            </configuration>
        </execution>
    </executions>
</plugin>
```

##### Steps to reproduce

<!-- unambiguous set of steps to reproduce the bug.-->

##### Related issues/PRs

<!-- has a similar issue/PR been reported/opened before? Please do a search in https://github.com/openapitools/openapi-generator/issues?utf8=%E2%9C%93&q=is%3Aissue%20 -->

##### Suggest a fix

<!-- if you can't fix the bug yourself, perhaps you can point to what might be
  causing the problem (line of code or commit), or simply make a suggestion -->

A fix could be to add a parser in eihter [https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/Java/JSON.mustache#L365](https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/Java/JSON.mustache#L365) or [https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/Java/libraries/okhttp-gson/JSON.mustache#L365](https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/Java/libraries/okhttp-gson/JSON.mustache#L365):

```
    /**
     * Gson TypeAdapter for JSR310 LocalDateTime type
     */
    public static class LocalDateTimeTypeAdapter extends TypeAdapter<LocalDateTime> {
        private DateTimeFormatter formatter;
        public LocalDateTimeTypeAdapter() {
            this(DateTimeFormatter.ISO_LOCAL_DATE_TIME);
        }

        public LocalDateTimeTypeAdapter(DateTimeFormatter formatter) {
            this.formatter = formatter;
        }

        public void setFormat(DateTimeFormatter dateFormat) {
            this.formatter = dateFormat;
        }

        @Override
        public void write(JsonWriter out, LocalDateTime date) throws IOException {
            if (date == null) {
                out.nullValue();
            } else {
                out.value(formatter.format(date));
            }
        }

        @Override
        public LocalDateTime read(JsonReader in) throws IOException {
            switch (in.peek()) {
                case NULL:
                    in.nextNull();
                    return null;
                default:
                    String date = in.nextString();
                    return LocalDateTime.parse(date, formatter);
            }
        }
    }
```

--- a/modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java
+++ b/modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java
@@ -674,9 +674,10 @@ public abstract class AbstractJavaCodegen extends DefaultCodegen implements Code
             importMapping.put("LocalDate", "java.time.LocalDate");
             typeMapping.put("time-local","LocalTime");
             importMapping.put("LocalTime", "java.time.LocalTime");
+            typeMapping.put("date-time-local", "LocalDateTime");
+            importMapping.put("LocalDateTime", "java.time.LocalDateTime");
             if ("java8-localdatetime".equals(dateLibrary)) {
                 typeMapping.put("DateTime", "LocalDateTime");
-                importMapping.put("LocalDateTime", "java.time.LocalDateTime");
             } else {
                 typeMapping.put("DateTime", "OffsetDateTime");
                 importMapping.put("OffsetDateTime", "java.time.OffsetDateTime");
@@ -1450,6 +1451,13 @@ public abstract class AbstractJavaCodegen extends DefaultCodegen implements Code
                 }
             }
             return null;
+        } else if (ModelUtils.isDateTimeLocalSchema(schema)) {
+            if (schema.getDefault() != null) {
+                if ("java8".equals(getDateLibrary())) {
+                    return String.format(Locale.ROOT, "LocalDateTime.parse(\"%s\")", String.valueOf(schema.getDefault()));
+                }
+            }
+            return null;
         } else if (ModelUtils.isStringSchema(schema)) {
             if (schema.getDefault() != null) {
                 String _default;
@@ -1535,6 +1543,10 @@ public abstract class AbstractJavaCodegen extends DefaultCodegen implements Code
                             } else if(ModelUtils.isTimeLocalSchema(propertySchema)) {
                                 if("java8".equals(getDateLibrary())) {
                                     defaultPropertyExpression = String.format(Locale.ROOT, "java.time.LocalTime.parse(\"%s\")", value.asText());
+                                }
+                            } else if(ModelUtils.isDateTimeLocalSchema(propertySchema)) {
+                                if("java8".equals(getDateLibrary())) {
+                                    defaultPropertyExpression = String.format(Locale.ROOT, "java.time.LocalDateTime.parse(\"%s\")", value.asText());
                                 }
                             } else if(ModelUtils.isUUIDSchema(propertySchema)) {
                                 defaultPropertyExpression = "java.util.UUID.fromString(\"" + value.asText() + "\")";

--- a/modules/openapi-generator/src/main/java/org/openapitools/codegen/utils/ModelUtils.java
+++ b/modules/openapi-generator/src/main/java/org/openapitools/codegen/utils/ModelUtils.java
@@ -715,6 +715,12 @@ public class ModelUtils {
                         && SchemaTypeUtil.DATE_TIME_FORMAT.equals(schema.getFormat()));
     }
 
+    public static boolean isDateTimeLocalSchema(Schema schema) {
+        // format: date-time-local, see https://spec.openapis.org/registry/format/date-time-local.html
+        return (SchemaTypeUtil.STRING_TYPE.equals(getType(schema))
+                        && "date-time-local".equals(schema.getFormat()));
+    }
+
     public static boolean isTimeLocalSchema(Schema schema) {
         // format: time-local, see https://spec.openapis.org/registry/format/time-local.html
         return (SchemaTypeUtil.STRING_TYPE.equals(getType(schema))

--- a/modules/openapi-generator/src/test/java/org/openapitools/codegen/java/AbstractJavaCodegenTest.java
+++ b/modules/openapi-generator/src/test/java/org/openapitools/codegen/java/AbstractJavaCodegenTest.java
@@ -537,6 +537,15 @@ public class AbstractJavaCodegenTest {
 
         // dateLibrary <> java8
         Assert.assertEquals(defaultValue, "10:15:30");
+
+        // Test default value for date-time-local format
+        StringSchema dateTimeLocalSchema = new StringSchema();
+        dateTimeLocalSchema.setFormat("date-time-local");
+        dateTimeLocalSchema.setDefault(LocalDateTime.parse("2007-12-03T10:15:30"));
+        defaultValue = codegen.toDefaultValue(dateTimeLocalSchema);
+
+        // dateLibrary <> java8
+        Assert.assertEquals(defaultValue, "2007-12-03T10:15:30");
     }
 
     @Test
@@ -606,6 +615,13 @@ public class AbstractJavaCodegenTest {
         timeLocalSchema.setDefault("10:15:30");
         defaultValue = codegen.toDefaultValue(codegen.fromProperty("", timeLocalSchema), timeLocalSchema);
         Assert.assertEquals(defaultValue, "LocalTime.parse(\"10:15:30\")");
+
+        // Test default value for date-time-local format
+        StringSchema dateTimeLocalSchema = new StringSchema();
+        dateTimeLocalSchema.setFormat("date-time-local");
+        dateTimeLocalSchema.setDefault("2007-12-03T10:15:30");
+        defaultValue = codegen.toDefaultValue(codegen.fromProperty("", dateTimeLocalSchema), dateTimeLocalSchema);
+        Assert.assertEquals(defaultValue, "LocalDateTime.parse(\"2007-12-03T10:15:30\")");
     }
 
     @Test

--- a/modules/openapi-generator/src/test/java/org/openapitools/codegen/java/spring/SpringCodegenTest.java
+++ b/modules/openapi-generator/src/test/java/org/openapitools/codegen/java/spring/SpringCodegenTest.java
@@ -413,6 +413,35 @@ public class SpringCodegenTest {
     }
 
     @Test
+    public void generateLocalDateTimeForDateTimeLocalFormat() throws IOException {
+        File output = Files.createTempDirectory("test").toFile().getCanonicalFile();
+        output.deleteOnExit();
+        String outputPath = output.getAbsolutePath().replace('\\', '/');
+
+        OpenAPI openAPI = new OpenAPIParser()
+                .readLocation("src/test/resources/3_0/spring/date-time-parameter-types-for-testing.yml", null, new ParseOptions()).getOpenAPI();
+
+        SpringCodegen codegen = new SpringCodegen();
+        codegen.setOutputDir(output.getAbsolutePath());
+
+        ClientOptInput input = new ClientOptInput();
+        input.openAPI(openAPI);
+        input.config(codegen);
+
+        DefaultGenerator generator = new DefaultGenerator();
+        generator.setGeneratorPropertyDefault(CodegenConstants.MODELS, "true");
+        generator.setGeneratorPropertyDefault(CodegenConstants.MODEL_TESTS, "false");
+        generator.setGeneratorPropertyDefault(CodegenConstants.MODEL_DOCS, "false");
+        generator.setGeneratorPropertyDefault(CodegenConstants.APIS, "false");
+        generator.setGenerateMetadata(false);
+        generator.opts(input).generate();
+
+        JavaFileAssert.assertThat(Paths.get(outputPath + "/src/main/java/org/openapitools/model/Pet.java"))
+                .hasImports("java.time.LocalDateTime")
+                .assertProperty("adoptionDate").withType("LocalDateTime");
+    }
+
+    @Test
     public void interfaceDefaultImplDisableWithResponseWrapper() {
         final SpringCodegen codegen = new SpringCodegen();
         codegen.additionalProperties().put(RESPONSE_WRAPPER, "aWrapper");

--- a/modules/openapi-generator/src/test/resources/3_0/spring/date-time-parameter-types-for-testing.yml
+++ b/modules/openapi-generator/src/test/resources/3_0/spring/date-time-parameter-types-for-testing.yml
@@ -106,3 +106,7 @@ components:
           type: string
           format: time-local
           default: '10:15:30'
+        adoptionDate:
+          type: string
+          format: date-time-local
+          default: '2007-12-03T10:15:30'

--- a/samples/client/petstore/spring-cloud-date-time/src/main/java/org/openapitools/model/Pet.java
+++ b/samples/client/petstore/spring-cloud-date-time/src/main/java/org/openapitools/model/Pet.java
@@ -6,6 +6,7 @@ import com.fasterxml.jackson.annotation.JsonProperty;
 import com.fasterxml.jackson.annotation.JsonCreator;
 import java.math.BigDecimal;
 import java.time.LocalDate;
+import java.time.LocalDateTime;
 import java.time.LocalTime;
 import java.time.OffsetDateTime;
 import org.springframework.format.annotation.DateTimeFormat;
@@ -42,6 +43,8 @@ public class Pet {
   private LocalDate dateOfBirth = LocalDate.parse("2021-01-01");
 
   private LocalTime feedingTime = LocalTime.parse("10:15:30");
+
+  private LocalDateTime adoptionDate = LocalDateTime.parse("2007-12-03T10:15:30");
 
   public Pet() {
     super();
@@ -201,6 +204,27 @@ public class Pet {
     this.feedingTime = feedingTime;
   }
 
+  public Pet adoptionDate(LocalDateTime adoptionDate) {
+    this.adoptionDate = adoptionDate;
+    return this;
+  }
+
+  /**
+   * Get adoptionDate
+   * @return adoptionDate
+   */
+  @Valid 
+  @Schema(name = "adoptionDate", requiredMode = Schema.RequiredMode.NOT_REQUIRED)
+  @JsonProperty("adoptionDate")
+  public LocalDateTime getAdoptionDate() {
+    return adoptionDate;
+  }
+
+  @JsonProperty("adoptionDate")
+  public void setAdoptionDate(LocalDateTime adoptionDate) {
+    this.adoptionDate = adoptionDate;
+  }
+
   @Override
   public boolean equals(Object o) {
     if (this == o) {
@@ -216,12 +240,13 @@ public class Pet {
         Objects.equals(this.price, pet.price) &&
         Objects.equals(this.lastFeed, pet.lastFeed) &&
         Objects.equals(this.dateOfBirth, pet.dateOfBirth) &&
-        Objects.equals(this.feedingTime, pet.feedingTime);
+        Objects.equals(this.feedingTime, pet.feedingTime) &&
+        Objects.equals(this.adoptionDate, pet.adoptionDate);
   }
 
   @Override
   public int hashCode() {
-    return Objects.hash(atType, age, happy, price, lastFeed, dateOfBirth, feedingTime);
+    return Objects.hash(atType, age, happy, price, lastFeed, dateOfBirth, feedingTime, adoptionDate);
   }
 
   @Override
@@ -235,6 +260,7 @@ public class Pet {
     sb.append("    lastFeed: ").append(toIndentedString(lastFeed)).append("\n");
     sb.append("    dateOfBirth: ").append(toIndentedString(dateOfBirth)).append("\n");
     sb.append("    feedingTime: ").append(toIndentedString(feedingTime)).append("\n");
+    sb.append("    adoptionDate: ").append(toIndentedString(adoptionDate)).append("\n");
     sb.append("}");
     return sb.toString();
   }

--- a/samples/openapi3/client/petstore/spring-cloud-date-time/src/main/java/org/openapitools/model/Pet.java
+++ b/samples/openapi3/client/petstore/spring-cloud-date-time/src/main/java/org/openapitools/model/Pet.java
@@ -6,6 +6,7 @@ import com.fasterxml.jackson.annotation.JsonProperty;
 import com.fasterxml.jackson.annotation.JsonCreator;
 import java.math.BigDecimal;
 import java.time.LocalDate;
+import java.time.LocalDateTime;
 import java.time.LocalTime;
 import java.time.OffsetDateTime;
 import org.springframework.format.annotation.DateTimeFormat;
@@ -42,6 +43,8 @@ public class Pet {
   private LocalDate dateOfBirth = LocalDate.parse("2021-01-01");
 
   private LocalTime feedingTime = LocalTime.parse("10:15:30");
+
+  private LocalDateTime adoptionDate = LocalDateTime.parse("2007-12-03T10:15:30");
 
   public Pet() {
     super();
@@ -201,6 +204,27 @@ public class Pet {
     this.feedingTime = feedingTime;
   }
 
+  public Pet adoptionDate(LocalDateTime adoptionDate) {
+    this.adoptionDate = adoptionDate;
+    return this;
+  }
+
+  /**
+   * Get adoptionDate
+   * @return adoptionDate
+   */
+  @Valid 
+  @Schema(name = "adoptionDate", requiredMode = Schema.RequiredMode.NOT_REQUIRED)
+  @JsonProperty("adoptionDate")
+  public LocalDateTime getAdoptionDate() {
+    return adoptionDate;
+  }
+
+  @JsonProperty("adoptionDate")
+  public void setAdoptionDate(LocalDateTime adoptionDate) {
+    this.adoptionDate = adoptionDate;
+  }
+
   @Override
   public boolean equals(Object o) {
     if (this == o) {
@@ -216,12 +240,13 @@ public class Pet {
         Objects.equals(this.price, pet.price) &&
         Objects.equals(this.lastFeed, pet.lastFeed) &&
         Objects.equals(this.dateOfBirth, pet.dateOfBirth) &&
-        Objects.equals(this.feedingTime, pet.feedingTime);
+        Objects.equals(this.feedingTime, pet.feedingTime) &&
+        Objects.equals(this.adoptionDate, pet.adoptionDate);
   }
 
   @Override
   public int hashCode() {
-    return Objects.hash(atType, age, happy, price, lastFeed, dateOfBirth, feedingTime);
+    return Objects.hash(atType, age, happy, price, lastFeed, dateOfBirth, feedingTime, adoptionDate);
   }
 
   @Override
@@ -235,6 +260,7 @@ public class Pet {
     sb.append("    lastFeed: ").append(toIndentedString(lastFeed)).append("\n");
     sb.append("    dateOfBirth: ").append(toIndentedString(dateOfBirth)).append("\n");
     sb.append("    feedingTime: ").append(toIndentedString(feedingTime)).append("\n");
+    sb.append("    adoptionDate: ").append(toIndentedString(adoptionDate)).append("\n");
     sb.append("}");
     return sb.toString();
   }

